enhancements/microshift/advertise-version-details.md

```diff
@@ -0,0 +1,272 @@
+on OCP and MicroShift. Ideally this is not needed, but in practice it
+is useful because of different APIs being available and different
+constraints being in place. It is possible for a cluster admin to read
+the ClusterVersion resource in an OCP cluster to discover that it is
+OCP and what version is running. This approach has two
+drawbacks. First, not all console users have cluster admin privileges,
+which means that the console cannot show them the cluster version
+details. Second, the ClusterVersion API is not present in a MicroShift
+deployment, so instead [we created a
+ConfigMap](https://github.com/openshift/microshift/pull/776) in the
+`openshift-public` namespace containing some basic version
+details. That means a caller must know two different APIs to check the
+version and product they are communicating with. This enhancement
+describes how we can standardize on one to make it easier for all
+authenticated users to obtain basic version details for a cluster.
+
+### User Stories
+
+As a console user, I want to see the version of OCP I am using so that
+I can refer to the correct version of the documentation and API
+references.
+
+As an authenticated API caller, I want to obtain the product name and
+version of the OpenShift variant I am communicating with so that I can
+use it to make decisions on the client side.
+
+### Goals
+
+* Avoid introducing a new CRD, so we do not increase the overhead for
+  MicroShift.
+* Use a standard Kubernetes type, to keep access easy for arbitrary
+  API clients.
+* Define a standard location for version information that all
+  OpenShift variants can support.
+
+### Non-Goals
+
+* This change does not introduce a new way to update the version of a
+  cluster.
+* This change does not introduce a new way for a client to discover
+  that a cluster is being upgraded.
+* This change does not introduce a new way for a client to discover
+  the cluster topology (single-node, HA, external control-plane, etc.).
+
+## Proposal
+
+Because we want arbitrary clients to be able to read and understand
+the data, we will use a standard Kubernetes ConfigMap.
+
+Because we want all OpenShift variants to publish the information in
+the same way and that the data should not be writable by API callers,
+we will specify that the ConfigMap should be named `version` and
+placed in the `openshift` namespace.
+
+Because we want all OpenShift variants to report the same type of
+information, we will keep the contents of the ConfigMap brief and
+specify the data key names in this enhancement:
+
+```yaml
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: version
+  namespace: openshift
+data:
+  product: ""
+  major: ""
+  minor: ""
+  version: ""
+```
+
+- `product` should hold a representation of the OpenShift variant
+  name. For example, `OCP` for OpenShift in standalone or external
+  control-plane configurations of all topologies and `MicroShift` for
+  MicroShift instances.
```

Comment:
the downside to doing this as a configmap and not a CRD w/ an enum field, is that it becomes hard for clients to know what values they are supposed to be looking for to compare against.

Reply:
Yeah, that's a good point. I don't know how many different values there are likely to ever be. I don't see us wanting to let people differentiate between standalone OpenShift and HyperShift this way, for example. But there could be more in the future.

How do people know what values to look for in `/etc/os-release`?

Reply:
> How do people know what values to look for in /etc/os-release?

i don't know, i also don't know how many applications code behavior based on it though

Reply:
Just to be clear, the version advertised is the control plane?  If there is skew between control plane members during upgrade, which value should be reported?  N or N+1?

Reply:
Never mind above, see it’s clarified below.

Reply:
> > How do people know what values to look for in /etc/os-release?
> 
> i don't know, i also don't know how many applications code behavior based on it though

Think about it more from the installer or operator perspective than the app itself. Lots of Ansible roles & playbooks have different actions for different platforms, for example. It appears to be the case that Helm charts make similar decisions, for example between *KS clusters and OpenShift clusters. I'll update the doc to highlight this use case more directly.

Reply:
Hmm, also, how do we handle things like OSD/ROSA that are layered on top of OCP (and which will at some point possibly be layered on top of HyperShift, etc.)?  Knowing that you cluster is OSD definitely impacts where users go for docs ([OSD][1] vs. [OCP][2]).  I'm less clear on whether console devs, authed API callers, or application devs would want to switch on OSD vs. OCP, etc.

[1]: https://docs.openshift.com/dedicated/welcome/index.html
[2]: https://docs.openshift.com/container-platform/4.10/welcome/index.html

Reply:
> I'm less clear on whether console devs, authed API callers, or application devs would want to switch on OSD vs. OCP, etc.

I don't think we want to differentiate between those products, but I wouldn't rule it out. If we decide we need to, we can figure out how to do that in another enhancement.

Reply:
Do we want to decide if `version` is squatting on a too-generic name, or if we have something that specifically collects things at the OCP/MicroShift/Hypershift level, but leaves room for OSD/ROSA/... to use their own managed ConfigMap alongside?  Or perhaps they'd use their own namespace, and all of `openshift` is ours to allow the generic `version` name?

Reply:
What attributes defines a variant  separate than OCP?  Do we agree that OCP is different than clusters on hypershift? There is also Single node OpenShift (SNO). . I am not worried about it as long as we have clear rules to decide a variant.

Reply:
From a workload's perspective we want standalone OpenShift and HyperShift to be the same. Same for SNO, ROSA, OSD, etc. They all have the platform operators, OLM, etc. and are all consumable in the same way, so to me they are all "OCP". MicroShift isn't consumed in the same way, so it has a different product name.

Reply:
I added a non-goal about not differentiating those other variations.